### DISCUSSION TITLE
How to use Lit with CoffeeScript?

### DISCUSSION THREAD

**Discussion by CedX** [lit/lit#4835] [General]
I'm trying to create a sample app using [CoffeeScript](https://coffeescript.org) instead of JS/TS.
On the whole, everything works fine except for one crucial thing: the reactive properties.

It seems that Lit doesn't recognize the way CoffeeScript generates the `properties` static member.
In CoffeeScript:

```coffeescript
export class MyElement extends LitElement
  @properties =
    myReactiveProperty: state: on
```

The generated JavaScript code:

```javascript
export var MyElement = (function() {
  class MyElement extends LitElement {};

  MyElement.properties = {
    myReactiveProperty: {state: true}
  };

  return MyElement;
}).call(this);
```

I've tried other solutions (e.g. `Object.defineProperty()`) without success...
Does anyone know how to generate the `properties` static member so that it is recognized by Lit?

**Comment by justinfagnani** [comment]
That seems like it should work. What's the error you're getting?

**Reply by CedX** [reply]
That's the problem: I have absolutely no errors.

It's just that reactive properties are ignored: they don't trigger any update/rendering when they are changed (and attributes use their default value defined in the component constructor, instead of the one specified in the HTML template).
When I log the `render()` method, it is called just once when the component is connected.

The object properties are indeed defined, but this is not reflected in the view.

**Reply by justinfagnani** [reply]
Can you make a reproduction in the lit.dev playground? When I made a class like you have above, it worked fine.

**Reply by CedX** [reply]
Obviously, there's a subtle bug in my CS code. With a bare component, it works. But not with this one:

https://lit.dev/playground/#project=W3sibmFtZSI6ImluZGV4Lmh0bWwiLCJjb250ZW50IjoiPCFET0NUWVBFIGh0bWw-XG48aGVhZD5cbiAgPHNjcmlwdCB0eXBlPVwibW9kdWxlXCIgc3JjPVwiLi90aGVtZS1kcm9wZG93bi5qc1wiPjwvc2NyaXB0PlxuPC9oZWFkPlxuPGJvZHk-XG4gIDx0aGVtZS1kcm9wZG93biBsYWJlbD1cIkhlbGxvXCI-PC90aGVtZS1kcm9wZG93bj5cbjwvYm9keT5cbiJ9LHsibmFtZSI6InBhY2thZ2UuanNvbiIsImNvbnRlbnQiOiJ7XG4gIFwiZGVwZW5kZW5jaWVzXCI6IHtcbiAgICBcImxpdFwiOiBcIl4zLjAuMFwiLFxuICAgIFwiQGxpdC9yZWFjdGl2ZS1lbGVtZW50XCI6IFwiXjIuMC4wXCIsXG4gICAgXCJsaXQtZWxlbWVudFwiOiBcIl40LjAuMFwiLFxuICAgIFwibGl0LWh0bWxcIjogXCJeMy4wLjBcIlxuICB9XG59IiwiaGlkZGVuIjp0cnVlfSx7Im5hbWUiOiJ0aGVtZS1kcm9wZG93bi5qcyIsImNvbnRlbnQiOiJpbXBvcnQge1xuICBodG1sLFxuICBMaXRFbGVtZW50XG59IGZyb20gXCJsaXRcIjtcblxuaW1wb3J0IHtcbiAgd2hlbiBhcyBfd2hlblxufSBmcm9tIFwibGl0L2RpcmVjdGl2ZXMvd2hlbi5qc1wiO1xuXG5leHBvcnQgdmFyIFRoZW1lID0gT2JqZWN0LmZyZWV6ZSh7XG4gIGxpZ2h0OiBcImxpZ2h0XCIsXG4gIGRhcms6IFwiZGFya1wiLFxuICBhdXRvOiBcImF1dG9cIlxufSk7XG5cbmV4cG9ydCB2YXIgVGhlbWVEcm9wZG93biA9IChmdW5jdGlvbigpIHtcbiAgY2xhc3MgVGhlbWVEcm9wZG93biBleHRlbmRzIExpdEVsZW1lbnQge1xuICAgIGNvbnN0cnVjdG9yKCkge1xuICAgICAgc3VwZXIoKTtcbiAgICAgIHRoaXMubGFiZWwgPSBcIlwiO1xuICAgICAgdGhpcy5fdGhlbWUgPSBUaGVtZS5hdXRvO1xuICAgIH1cblxuICAgIHJlbmRlcigpIHtcbiAgICAgIGNvbnNvbGUubG9nKFwicmVuZGVyXCIpO1xuICAgICAgcmV0dXJuIGh0bWxgXG4gICAgICAgIDxwPlxuXHQgICAgICAke193aGVuKHRoaXMubGFiZWwsICgpID0-IGh0bWxgJHt0aGlzLmxhYmVsfWApfVxuXHQgICAgICA8Yj4ke3RoaXMuX3RoZW1lfTwvYj5cbiAgICAgICAgPC9wPlxuICAgICAgICA8dWw-XG5cdCAgICAgICR7T2JqZWN0LnZhbHVlcyhUaGVtZSkubWFwKCh2YWx1ZSkgPT4gaHRtbGBcbiAgICAgICAgICAgIDxsaT5cblx0ICAgICAgICAgIDxidXR0b24gQGNsaWNrPSR7KCkgPT4gdGhpcy5zZXRUaGVtZSh2YWx1ZSl9PlxuICAgICAgXHRcdCAgICAke3ZhbHVlfVxuXHRcdCAgICAgICAgJHtfd2hlbih2YWx1ZSA9PT0gdGhpcy5fdGhlbWUsICgpID0-IGh0bWxgKGN1cnJlbnQpYCl9XG5cdCAgICAgICAgICA8L2J1dHRvbj5cbiAgICAgICAgICAgIDwvbGk-YFxuICAgICAgICAgICl9XG4gICAgICAgIDwvdWw-YDtcbiAgICB9XG5cbiAgICBzZXRUaGVtZSh2YWx1ZSkge1xuICAgICAgdGhpcy5fdGhlbWUgPSB2YWx1ZTtcbiAgICAgIGNvbnNvbGUubG9nKFwidGhpcy5fdGhlbWVcIiwgdmFsdWUpO1xuICAgIH1cblxuICB9O1xuXG4gIGN1c3RvbUVsZW1lbnRzLmRlZmluZShcInRoZW1lLWRyb3Bkb3duXCIsIFRoZW1lRHJvcGRvd24pO1xuXG4gIFRoZW1lRHJvcGRvd24ucHJvcGVydGllcyA9IHtcbiAgICBsYWJlbDoge1xuICAgICAgdHlwZTogU3RyaW5nXG4gICAgfSxcbiAgICBfdGhlbWU6IHtcbiAgICAgIHN0YXRlOiB0cnVlXG4gICAgfVxuICB9O1xuXG4gIHJldHVybiBUaGVtZURyb3Bkb3duO1xuXG59KS5jYWxsKHRoaXMpO1xuIn1d

**Reply by CedX** [reply]
Bingo: (not so subtle) the call to `customElements.define()` was made before defining the reactive properties!
CoffeeScript lets you place arbitrary code in the body of a class (i.e. outside any method): this code will be added after the class in the generated JavaScript code.
But I made the mistake of placing the `customElements.define()` call at the beginning of the class body, before the `properties` member... 😿

**Comment by CedX** [comment]
In short: Lit works perfectly with CoffeeScript 😸 
It was just a bug in my code.

**Reply by CedX** [reply]
For the curious: an example of a CoffeeScript component.
https://github.com/cedx/core/blob/main/src/html/ui/theme_dropdown.coffee
